### PR TITLE
Fixes Radiation diagnostic bug and refactors modal_aero_calcsize

### PR DESCRIPTION
**Pull request by singhbalwinder**
Fixes a bug where radiation diagnostic call with the same aerosol species 
as radiation climate call was producing NBFB answers with respect to
radiation climate call. The bug is fixed by eliminating calls to
modal_aero_calcsize_diag subroutine in modal_aer_opt.F90. This subroutine
was called for radiation diagnostic calls. Now, modal_aero_calcsize_sub is
called for both diagnostic and prognostic calls.

I have refactored modal_aero_calcsize_sub subroutine extensively, so that
it is easy to understand different processes taking place in this subroutine.
While doing the refactor, I found another bug which was causing the model to
use incorrect log of sigma g (geometric std. dev. of aerosol mode) during 
accumulation<->aitken transfer.

I have also modified code to avoid "allocation" and "deallocation" of arrays 
in a loop so that multiple radiation diagnostic calls do not run out of
memory.

This PR is NBFB as both prognostic and diagnostic calls for calculating
aerosol sizes and water uptake are now using the current "state" (mass
mixing ratios of the species). Earlier, an older "state" was used to compute 
these quantities.

Fixes: #3468
Fixes: #2575
[NBFB]

### CODE REVIEW

**Comment by singhbalwinder** [comment]
@ndkeen : I have rebased this branch. Please feel free to test it to see if it resolves the memory issue.

**Comment by golaz** [comment]
@rljacob : this is an important fix to the diagnostic radiation feature. A could imagine a number of v2 users wanting to turn on this feature. We should try to get this into v2 if possible.

**Comment by rljacob** [comment]
@golaz if you want it in v2, it can be in v2.

File: components/eam/src/chemistry/utils/modal_aero_wateruptake.F90 [hunk 1]
```diff
@@ -336,15 +370,15 @@ subroutine modal_aero_wateruptake_dr(state, pbuf, list_idx_in, dgnumdry_m, dgnum
...
-   end do ! m = 1, nmodes
+   end do    ! modes
 
    !----------------------------------------------------------------------------
    ! specify clear air relative humidity
 
    if (present(clear_rh_in)) then
 
-      ! use input relative humidity
-      rh(1:ncol,1:pver) = clear_rh_in(1:ncol,1:pver)
```
Comment (whannah1) [inline components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:347]: Wait, what's going on here? Are you undoing this change I added for the MMF?

File: components/eam/src/chemistry/modal_aero/aero_model.F90 [hunk 2]
```diff
@@ -1529,7 +1529,7 @@ subroutine aero_model_wetdep(dt, dlf, dlf2, cmfmc2, state,                    &!
     ! for prognostic modal aerosols the transfer of mass between aitken and 
     ! accumulation modes is done in conjunction with the dry radius calculation
     call t_startf('calcsize')
-    call modal_aero_calcsize_sub(state, ptend, dt, pbuf)
+    call modal_aero_calcsize_sub(state, pbuf, dt, ptend)
```
Comment (whannah1) [inline components/eam/src/chemistry/modal_aero/aero_model.F90:1532]: this is going to break the MMF unless the corresponding change is made in src/physics/crm/physpkg.F90

File: components/eam/src/physics/rrtmg/radiation.F90 [hunk 3]
```diff
@@ -850,7 +850,7 @@ subroutine radiation_tend(state,ptend, pbuf, &
 
     ! Arguments
     logical,  intent(in)    :: is_cmip6_volc    ! true if cmip6 style volcanic file is read otherwise false 
-    real(r8), intent(in)    :: landfrac(pcols)  ! land fraction
+    real(r8), intent(in)    :: landfrac(pcols),dt  ! land fraction
```
Comment (whannah1) [inline components/eam/src/physics/rrtmg/radiation.F90:853]: Why isn't "dt" added on it's own line with a proper comment?

File: components/eam/src/chemistry/utils/modal_aero_wateruptake.F90 (hunk 1 above)
Comment (singhbalwinder) [reply components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:347]: Thanks @whannah1 for catching this. It got removed during cam->eam conflict resolution. I will reinstate this code and revisit this code again to ensure that there aren't anymore surprises.

File: components/eam/src/chemistry/modal_aero/aero_model.F90 (hunk 2 above)
Comment (singhbalwinder) [reply components/eam/src/chemistry/modal_aero/aero_model.F90:1532]: I am aware of that. I will revisit all the calls to calcsize and fix this order. I am making `ptend` optional, so I had to switch the order.

File: components/eam/src/physics/rrtmg/radiation.F90 (hunk 3 above)
Comment (singhbalwinder) [reply components/eam/src/physics/rrtmg/radiation.F90:853]: thanks. I will fix this as well.

File: components/eam/src/chemistry/modal_aero/aero_model.F90 (hunk 2 above)
Comment (whannah1) [reply components/eam/src/chemistry/modal_aero/aero_model.F90:1532]: ok, that makes sense.

File: components/eam/src/chemistry/utils/modal_aero_wateruptake.F90 (hunk 1 above)
Comment (whannah1) [reply components/eam/src/chemistry/utils/modal_aero_wateruptake.F90:347]: ah, I see, that's fine then.

**Comment by singhbalwinder** [comment]
I have fixed code to resolve @whannah1 comments. 
@whannah1 : Thanks for your quick review. My eventual goal is to remove calcsize_diag and use calcsize_sub for all dry size calculations. I will let you know when I make that change (either in this or a future PR).

**Comment by whannah1** [comment]
@singhbalwinder Thanks for addressing these changes. You might want to run a SMS test with the F-MMF2 compset to double check that the MMF is not affected. 

I'm happy to help with the future changes to the calcsize routines.

File: components/eam/src/physics/cam/aer_rad_props.F90 [hunk 4]
```diff
@@ -418,7 +420,7 @@ subroutine aer_rad_props_lw(is_cmip6_volc, list_idx, state, pbuf,  odap_aer)
 
    ! Contributions from modal aerosols.
    if (nmodes > 0) then
-      call modal_aero_lw(list_idx, state, pbuf, odap_aer)
+      call modal_aero_lw(list_idx, dt, state, pbuf, odap_aer)
```
Comment (wlin7) [inline components/eam/src/physics/cam/aer_rad_props.F90:423]: The addition of dummy variable "dt" does break the compilation of rrtmgp/cam_optics.F90 in its call to aer_rad_props_sw at line 506

```
call aer_rad_props_sw(icall, state, pbuf, &
                            count(night_indices > 0), night_indices, is_cmip6_volc, &
                            tau, tau_w, tau_w_g, tau_w_f)
```
Similar changes as in rrtmg/radiation.F90 is needed.

**Comment by wlin7** [comment]
> @singhbalwinder Thanks for addressing these changes. You might want to run a SMS test with the F-MMF2 compset to double check that the MMF is not affected.

Good call, @whannah1 ,  F-MMF2 test failed during compilation of rrtmgp/cam_optics.F90, as commented  by the code changes.

File: components/eam/src/physics/cam/aer_rad_props.F90 (hunk 4 above)
Comment (singhbalwinder) [reply components/eam/src/physics/cam/aer_rad_props.F90:423]: Thanks @wlin7 for your quick review. I haven't touched rrtmgp codes yet as I will have to go through the rrtmgp submodule for making those changes. I  am actively working on this PR. It touches a lot of interdependent codes so I will do rigorous testing before it is ready for merging to master.

In you last line, you mean "rrtmg**p**", right? "rrtmg" codes looks okay to me.

Comment (wlin7) [reply components/eam/src/physics/cam/aer_rad_props.F90:423]: @singhbalwinder , in that last line, I did mean rrtmg as a reference. Yes, the rrtmg version is ok. I meant to make a change to rrtmgp/cam_optics.F90 similar to what you did in rrtmg/radiation.F90. Sorry for the confusion.

**Comment by singhbalwinder** [comment]
I am closing this PR and will issue a fresh PR which will include these and some more changes related to this feature.

**Comment by rljacob** [comment]
Replaced partly by #3885

**Pull request closed without merging**